### ISSUE TITLE
SourceFolderProviders can't set test folders

### ISSUE THREAD

**Issue by kamildoleglo** [bazelbuild/intellij#3670]
If a test folder is not conforming to any regex in the project view file but is marked as a test source by the SourceFolderProvider, then the `setSourceFolderForLocation` method is called on the relevant SourceFolderProvider. However, if the provider then adds the folder as a test source anyway, the ContentEntryEditor removes it in the subsequent step, because the ContentEntry map has the (folder, isTestSource) pair as the unique key. This prevents the SourceFolderProviders from setting the test source folders altogether

**Comment by mai93** [comment]
Hello @kamildoleglo can you provide us with repro steps to better understand the issue and the fix you are proposing in https://github.com/bazelbuild/intellij/pull/3671?

**Comment by kamildoleglo** [comment]
Yeah, sorry, got a lot of work lately. I'll try to keep the discussion for this issue and the two PRs (https://github.com/bazelbuild/intellij/pull/3671/, https://github.com/bazelbuild/intellij/pull/3832/) here.
> Can you provide us with repro steps to better understand the issue and the fix you are proposing (...)

Just try to create any custom `SourceFolderProvider ` and mark directories as test ones. My use case is that I'm writing a Python language support, we use custom Python rules internally and the current solution (with `GenericFolderProvider`) doesn't work well. I thus created a separate plugin and plugged into some extension points provided by this Bazel IntelliJ plugin. I then created a custom `PythonSourceFolderProvider` that marks source and test directories. While the source directory marking works well, the test directory marking does not work as expected (explained a bit in my first comment). 

Upon further inspection and a small refactoring (see https://github.com/bazelbuild/intellij/pull/3832) it turns out that there is a way for me to make the test directories marking work (in most cases at least) by marking the folder's parent as a test source root, since the `contentEntry.removeSourceFolder(current)` is not called in the `if` branch for the parent. 
TBF the whole logic there is a bit unclear to me, I don't know why someone wrote it like that and it's not visible in the Git blame, since it has been copied from a private repo. I guess you could try to find that out if you have the access. 

Anyway, the second PR (https://github.com/bazelbuild/intellij/pull/3671/) fixes the `if` as explained in the description, however it may not be backward-compatible (well it actually isn't but I don't think someone is relying on that part of the code). If we don't want to break existing production code though, we could create an interface (or even extract the whole method to the interface) with a default implementation (the current one) thus allowing users to completely override this part of code (which seems more or less useless, since it tries to fix a difference between results got from `SourceFolderProviders` and the projectview file). 

The last question that remains is - why is this method written like that and can we just rewrite it entirely? I don't have the knowledge nor the access to the original repo to answer that

**Comment by raoarun** [comment]
@kamildoleglo thanks.

@mai93 anything else needed from @kamildoleglo ?

**Comment by mai93** [comment]
@tpasternak has reviewed the PRs mentioned in the issue, https://github.com/bazelbuild/intellij/pull/3832 is merged and https://github.com/bazelbuild/intellij/pull/3671 needs to be rebased.
